src/fenics_constitutive/solver.py

```diff
@@ -0,0 +1,318 @@
+                    self._tangent.append(df.fem.Function(QT_subspace))
+
+                    # subspaces for history
+                    history_0 = build_history(law, submesh, q_degree)
+                    history_1 = (
+                        {key: fn.copy() for key, fn in history_0.items()}
+                        if isinstance(history_0, dict)
+                        else history_0
+                    )
+                    self._history_0.append(history_0)
+                    self._history_1.append(history_1)
+            else:
+                law, cells = laws[0]
+                self.laws.append((law, cells))
+
+                # subspace for grad u
+                Q_grad_u_space = df.fem.FunctionSpace(mesh, Q_grad_u_e)
+                self._del_grad_u.append(df.fem.Function(Q_grad_u_space))
+
+                # Spaces for history
+                history_0 = build_history(law, mesh, q_degree)
+                history_1 = (
+                    {key: fn.copy() for key, fn in history_0.items()}
+                    if isinstance(history_0, dict)
+                    else history_0
+                )
+                self._history_0.append(history_0)
+                self._history_1.append(history_1)
+
+        self.stress_0 = df.fem.Function(QV)
+        self.stress_1 = df.fem.Function(QV)
+        self.tangent = df.fem.Function(QT)
+
+        u_, du = ufl.TestFunction(u.function_space), ufl.TrialFunction(u.function_space)
+
+        self.metadata = {"quadrature_degree": q_degree, "quadrature_scheme": "default"}
+        self.dxm = ufl.dx(metadata=self.metadata)
+
+        self.R_form = (
+            ufl.inner(ufl_mandel_strain(u_, constraint), self.stress_1) * self.dxm
+        )
+        self.dR_form = (
+            ufl.inner(
+                ufl_mandel_strain(du, constraint),
+                ufl.dot(self.tangent, ufl_mandel_strain(u_, constraint)),
+            )
+            * self.dxm
+        )
+
+        self._u = u
+        self._u0 = u.copy()
+        self._bcs = bcs
+        self._form_compiler_options = form_compiler_options
+        self._jit_options = jit_options
+
+        basix_celltype = getattr(basix.CellType, mesh.topology.cell_type.name)
+        self.q_points, _ = basix.make_quadrature(basix_celltype, q_degree)
+
+        self.del_grad_u_expr = df.fem.Expression(
+            ufl.nabla_grad(self._u - self._u0), self.q_points
+        )
+
+        # ### Note on JIT compilation of UFL forms
+        # if super().__init__(R, u, bcs, dR) is called within ElasticityProblem.__init__
+        # the user cannot add Neumann BCs.
+        # Therefore, the compilation (i.e. call to super().__init__()) is done when
+        # df.nls.petsc.NewtonSolver is initialized.
+        # df.nls.petsc.NewtonSolver will call
+        # self._A = fem.petsc.create_matrix(problem.a) and hence it would suffice
+        # to override the property ``a`` of NonlinearProblem.
```

Comment:
Do we want to keep that note?
Maybe better move it to the body of the property ``a`` and also update. The text refers to ElasticityProblem.

Reply:
Will take care of it when I write the docstrings